lib/read/lib/Cardano/Wallet/Read/Tx/Gen/Shelley.hs

```diff
@@ -0,0 +1,202 @@
+import Cardano.Ledger.Credential
+    ( Credential (KeyHashObj)
+    , PaymentCredential
+    , StakeReference (..)
+    )
+import Cardano.Ledger.Keys
+    ( KeyHash (..)
+    )
+import Cardano.Ledger.SafeHash
+    ( unsafeMakeSafeHash
+    )
+import Cardano.Ledger.Shelley.PParams
+    ( Update
+    )
+import Cardano.Ledger.Shelley.Tx
+    ( ShelleyTx
+    , pattern ShelleyTx
+    , ShelleyTxBody
+    , pattern ShelleyTxBody
+    , ShelleyTxOut
+    , pattern ShelleyTxOut
+    , TxIn (..)
+    )
+import Cardano.Ledger.Shelley.TxAuxData
+    ( ShelleyTxAuxData
+    )
+import Cardano.Ledger.Shelley.TxBody
+    ( DCert
+    , Withdrawals (Withdrawals)
+    )
+import Cardano.Ledger.Shelley.TxWits
+    ( ShelleyTxWits
+    )
+import Cardano.Ledger.Slot
+    ( SlotNo (..)
+    )
+import Cardano.Wallet.Read.Tx.Gen.TxParameters
+    ( Address (..)
+    , Index (..)
+    , Lovelace (..)
+    , TxId (..)
+    , TxParameters (..)
+    , exampleTxParameters
+    )
+import Data.ByteString
+    ( ByteString
+    )
+import Data.Foldable
+    ( toList
+    )
+import Data.List.NonEmpty
+    ( NonEmpty
+    )
+import Data.Maybe.Strict
+    ( StrictMaybe
+    , maybeToStrictMaybe
+    )
+import Data.Sequence.Strict
+    ( StrictSeq
+    , fromList
+    )
+import Data.Set
+    ( Set
+    )
+
+import qualified Cardano.Ledger.Core as L
+import qualified Cardano.Ledger.TxIn as L
+import qualified Data.ByteString.Short as B
+import qualified Data.Set as Set
+import GHC.Stack
+    ( HasCallStack
+    )
+
+mkShelleyTx
+    :: TxParameters
+    -> ShelleyTx (ShelleyEra StandardCrypto)
+mkShelleyTx TxParameters{txInputs, txOutputs} =
+    ShelleyTx (body txInputs txOutputs) wits aux
+
+aux :: StrictMaybe (ShelleyTxAuxData (ShelleyEra StandardCrypto))
+aux = maybeToStrictMaybe Nothing
+
+wits :: L.EraScript (era StandardCrypto) => ShelleyTxWits (era StandardCrypto)
+wits = mempty
+
+body
+    :: HasCallStack
+    => NonEmpty (Index, TxId)
+    -> NonEmpty (Address, Lovelace)
+    -> ShelleyTxBody (ShelleyEra StandardCrypto)
+body ins outs =
+    ShelleyTxBody
+        (txins ins)
+        (txouts outs)
+        certs
+        wdrls
+        txfee
+        slot
+        upd
+        auxb
+
+auxb :: StrictMaybe (AuxiliaryDataHash StandardCrypto)
+auxb = maybeToStrictMaybe Nothing
+
+upd :: StrictMaybe (Update era)
+upd = maybeToStrictMaybe Nothing
+
+slot :: SlotNo
+slot = SlotNo 0
+
+txfee :: Coin
+txfee = Coin 0
+
+wdrls :: Withdrawals StandardCrypto
+wdrls = Withdrawals mempty
+
+certs :: StrictSeq (DCert StandardCrypto)
+certs = mempty
+
+txouts
+    :: ( Era era
+       , L.Value era ~ Coin
+       , L.EraCrypto era ~ StandardCrypto
+       )
+    => NonEmpty (Address, Lovelace)
+    -> StrictSeq (ShelleyTxOut era)
+txouts xs = fromList $ do
+    (Address addr, Lovelace val) <- toList xs
+    pure $ ShelleyTxOut (mkShelleyPaymentPart addr) $ Coin val
+
+mkShelleyPaymentPart :: ByteString -> Addr StandardCrypto
+mkShelleyPaymentPart x = Addr Mainnet (payment x) staking
```

Comment:
👍 This name better reflects the purpose.

However, the problem I was trying to highlight partially remains. Specifically,

```
Cardano.Ledger.Binary.serialize (mkByronAddrFromXPub x)
  ≠
Cardano.Ledger.Address.serialiseAddr (mkShelleyPaymentPart x)
```

This means that in the following example

```
    byron 10
    tx3 <- tx $ do
        input (Index 0) tx2
        output (addrid 'f') (Lovelace 300) -- index 0
    shelley 20
    tx4 <- $ tx $ do
        input (Index 0) tx1 -- use tx1 (500)
        input (Index 1) tx3 -- use tx3 (200)
        output (addrid 'f') (Lovelace 700) -- index 0
```

the transactions `tx3` and `tx4` will send funds to **different** addresses, even though both are denoted by `addrid 'f'`.